### 1. Load Model
- Gemma 1B

In [1]:
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer

model_name = "google/gemma-3-1b-it"
device = torch.device("cuda" if torch.cuda.is_available() else "mps" if torch.backends.mps.is_available() else "cpu")

print(f"Using device: {device}")
print(f"Model name: {model_name}")

model = AutoModelForCausalLM.from_pretrained(model_name).to(device)
tokenizer = AutoTokenizer.from_pretrained(model_name)

/Users/vijay/code/rft-grpo/.venv/lib/python3.11/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


Using device: mps
Model name: google/gemma-3-1b-it


In [2]:
import numpy as np

def generate_response(prompt, num_return_sequences=2):
    inputs = tokenizer(prompt, return_tensors="pt").to(device)
    outputs = model.generate(
        **inputs, 
        max_new_tokens=10,
        temperature=0.5,      # sample generations
        return_dict_in_generate=True,
        output_scores=True,
        num_return_sequences=num_return_sequences,
    )
    transition_scores = model.compute_transition_scores(
        outputs.sequences, outputs.scores, normalize_logits=True
    )

    input_length = inputs.input_ids.shape[1]
    generated_tokens = outputs.sequences[:, input_length:]
    for i in range(generated_tokens.shape[0]):
        sum_logits = 0
        for tok, score in zip(generated_tokens[i], transition_scores[i]):
            # | token | token string | logits | probability
            score = score.cpu().numpy()
            sum_logits += score
            print(f"| {tok:5d} | {tokenizer.decode(tok):8s} | {score:.4f} | {np.exp(score):.2%}")

        response = tokenizer.decode(generated_tokens[i], skip_special_tokens=True)
        print(f"Response: {response}")
        print(f"Probability of the response: {np.exp(sum_logits):.2%}")
    
    return response

print(generate_response("Tell me a joke."))


|   108 | 

       | 0.0000 | 100.00%
| 11355 | Why      | 0.0000 | 100.00%
|  1602 |  did     | 0.0000 | 100.00%
|   506 |  the     | 0.0000 | 100.00%
| 30979 |  bicycle | -1.0309 | 35.67%
|  3798 |  fall    | 0.0000 | 100.00%
|  1024 |  over    | 0.0000 | 100.00%
| 236881 | ?        | 0.0000 | 100.00%
|   108 | 

       | -0.0537 | 94.77%
| 17574 | Because  | 0.0000 | 100.00%
Response: 

Why did the bicycle fall over?

Because
Probability of the response: 33.80%
|   108 | 

       | 0.0000 | 100.00%
| 11355 | Why      | 0.0000 | 100.00%
|  1602 |  did     | 0.0000 | 100.00%
|   506 |  the     | 0.0000 | 100.00%
| 30979 |  bicycle | -1.0309 | 35.67%
|  3798 |  fall    | 0.0000 | 100.00%
|  1024 |  over    | 0.0000 | 100.00%
| 236881 | ?        | 0.0000 | 100.00%
|   108 | 

       | -0.0537 | 94.77%
| 17574 | Because  | 0.0000 | 100.00%
Response: 

Why did the bicycle fall over?

Because
Probability of the response: 33.80%


Why did the bicycle fall over?

Because


In [93]:
from transformers.generation.configuration_utils import GenerationConfig

gc = GenerationConfig(
    do_sample=True
)

In [452]:
sample_input = "Tell me a joke."
input_pt = tokenizer(sample_input, return_tensors="pt").to(device)
input_len = input_pt.input_ids.shape[1]

sample_seqs =model.generate(
    **input_pt,
    max_new_tokens=10,
    temperature=0.5,
    return_dict_in_generate=True,
    output_scores=True,
    num_return_sequences=1,
)

transition_scores = model.compute_transition_scores(
    sample_seqs.sequences, sample_seqs.scores, normalize_logits=True
)

for i in range(sample_seqs.sequences.shape[0]):
    print(f"Sequence {i+1}: {tokenizer.decode(sample_seqs.sequences[i], skip_special_tokens=True)}")
    print(f"Probability: {torch.exp(transition_scores[i].sum()):.2%}")
    print()


Sequence 1: Tell me a joke.

Why did the bicycle fall over?

Because
Probability: 33.80%



In [462]:
for i in range(sample_seqs.sequences.shape[0]):  # no. of sequences
    sum_log_prob = 0
    seq_str = ""
    for j in range(sample_seqs.sequences[i].shape[0]-input_len):  # output time step
        # max_idx = torch.argmax(sample_seqs.scores[j][i])  # list of score per time stamp
        max_idx = sample_seqs.sequences[i][j+input_len] # sampled token
        log_probs = torch.log_softmax(sample_seqs.scores[j][i], dim=-1)
        sum_log_prob += log_probs[max_idx]
        gen_token = tokenizer.decode(max_idx)
        seq_str += gen_token
        print(f"\"{gen_token}\"\t| token {max_idx} prob: {torch.exp(log_probs[max_idx]):.2%} | log prob: {log_probs[max_idx]:.4}")
    print(f"Sequence {i+1} log prob: {sum_log_prob:.4}")
    print(f"Sequence {i+1} prob: {torch.exp(sum_log_prob):.2%}")
    print(f"Sequence {i+1} string: {seq_str}")
    print("-"*100)





"

"	| token 108 prob: 100.00% | log prob: 0.0
"Why"	| token 11355 prob: 100.00% | log prob: 0.0
" did"	| token 1602 prob: 100.00% | log prob: 0.0
" the"	| token 506 prob: 100.00% | log prob: 0.0
" bicycle"	| token 30979 prob: 35.67% | log prob: -1.031
" fall"	| token 3798 prob: 100.00% | log prob: 0.0
" over"	| token 1024 prob: 100.00% | log prob: 0.0
"?"	| token 236881 prob: 100.00% | log prob: 0.0
"

"	| token 108 prob: 94.77% | log prob: -0.0537
"Because"	| token 17574 prob: 100.00% | log prob: 0.0
Sequence 1 log prob: -1.085
Sequence 1 prob: 33.80%
Sequence 1 string: 

Why did the bicycle fall over?

Because
----------------------------------------------------------------------------------------------------


-----------

### 2. Dataset
- Wordle

In [4]:
from datasets import load_dataset

dataset_name = "predibase/wordle-grpo"
dataset = load_dataset(dataset_name, split="train")

print(dataset)

Generating train split: 100%|██████████| 76/76 [00:00<00:00, 6558.45 examples/s]

Dataset({
    features: ['prompt', 'word_list', 'past_guess_history', 'secret'],
    num_rows: 76
})


In [20]:
sample = dataset[1]
print(sample['past_guess_history'])
print(sample['secret'])
print(sample['prompt'])


[['CRANE', 'C(x) R(x) A(-) N(x) E(-)'], ['SWEAT', 'S(x) W(x) E(-) A(-) T(x)']]
ALLEY
<|im_start|>system

You are playing Wordle, a word-guessing game.

### Game Rules:
- You have **6 tries** to guess a secret **5-letter** word.
- Each guess must be a valid **5-letter English word**.
- After each guess, you will receive feedback indicating how close your guess was.

### Feedback Format:
Each letter in your guess will receive one of three symbols:
1. ✓ : The letter is in the word and in the CORRECT position.
2. - : The letter is in the word but in the WRONG position.
3. x : The letter is NOT in the word.

### Example:
Secret Word: BRISK

Guess 1: STORM → Feedback: S(-) T(x) O(x) R(-) M(x)
Guess 2: BRAVE → Feedback: B(✓) R(✓) A(x) V(x) E(x)
Guess 3: BRISK → Feedback: B(✓) R(✓) I(✓) S(✓) K(✓)

### Response Format:
Think through the problem and feedback step by step. Make sure to first add your step by step thought process within <think> </think> tags. Then, return your guessed word in the 

In [32]:
res = generate_response(sample['prompt'])

print(res)

<|im_start|>system

You are playing Wordle, a word-guessing game.

### Game Rules:
- You have **6 tries** to guess a secret **5-letter** word.
- Each guess must be a valid **5-letter English word**.
- After each guess, you will receive feedback indicating how close your guess was.

### Feedback Format:
Each letter in your guess will receive one of three symbols:
1. ✓ : The letter is in the word and in the CORRECT position.
2. - : The letter is in the word but in the WRONG position.
3. x : The letter is NOT in the word.

### Example:
Secret Word: BRISK

Guess 1: STORM → Feedback: S(-) T(x) O(x) R(-) M(x)
Guess 2: BRAVE → Feedback: B(✓) R(✓) A(x) V(x) E(x)
Guess 3: BRISK → Feedback: B(✓) R(✓) I(✓) S(✓) K(✓)

### Response Format:
Think through the problem and feedback step by step. Make sure to first add your step by step thought process within <think> </think> tags. Then, return your guessed word in the following format: <guess> guessed-word </guess>.
<|im_end|>
<|im_start|>user
Make a n

In [31]:
messages = [
    {"role": "system", "content": "You are a helpful assistant."},
    {"role": "user", "content": "What is the capital of France?"},
    {"role": "assistant", "content": "It's Paris!"},
    {"role": "user", "content": "and India?"},
]
text = tokenizer.apply_chat_template(messages, tokenize=False, add_generation_prompt=True)
print(text)

<bos><start_of_turn>user
You are a helpful assistant.

What is the capital of France?<end_of_turn>
<start_of_turn>model
It's Paris!<end_of_turn>
<start_of_turn>user
and India?<end_of_turn>
<start_of_turn>model



In [37]:
def extract_guess_from_text(text):
    """
    Extracts the guess from between <guess> and </guess> tags in the input text.
    There should be only one pair of <guess>...</guess> tags.
    Returns the guess as a string, stripped of whitespace.
    Raises ValueError if the tags are missing or if there are multiple pairs.
    """
    import re
    matches = re.findall(r'<guess>(.*?)</guess>', text, re.DOTALL)
    if len(matches) != 1:
        raise ValueError(f"Expected exactly one <guess>...</guess> tag, found {len(matches)}.")
    return matches[0].strip()


ttt = """<think> some dumb thinking </think>
<guess> HELLO </guess>
"""
extract_guess_from_text(ttt)


'HELLO'

-------------------

### 3. Reward function

- Reward function for wordle


In [3]:
def reward_wordle(y_true: str, y_pred: str) -> float:
    """
    Reward function to calculate reward for wordle guess.
    Heuristic:
    if y_pred is invalid (!=5 or non-alpha) : 0
    for each position correct : 0.1 x n
    for each alpha correct in wrong position: 0.05 x n
    if extact match : 1
    """
    y_true = y_true.strip().upper()
    y_pred = y_pred.strip().upper()
    # Check validity
    if len(y_pred) != 5 or not y_pred.isalpha():
        return 0.0
    if y_pred == y_true:
        return 1.0
    reward = 0.0
    # Count correct positions
    for i in range(5):
        if y_pred[i] == y_true[i]:
            reward += 0.1
    # Count correct letters in wrong positions
    # To avoid double-counting, mark matched positions
    true_counts = {}
    pred_counts = {}
    for i in range(5):
        if y_pred[i] != y_true[i]:
            true_counts[y_true[i]] = true_counts.get(y_true[i], 0) + 1
            pred_counts[y_pred[i]] = pred_counts.get(y_pred[i], 0) + 1
    for letter in pred_counts:
        if letter in true_counts:
            reward += 0.05 * min(pred_counts[letter], true_counts[letter])
    return reward


In [4]:
print(reward_wordle("CRANE", "CRANE"))  # exact match, expect 1.0
print(reward_wordle("CRANE", "CRONY"))  # 3 correct positions (C, R, N), expect 0.3
print(reward_wordle("CRANE", "CANER"))  # 1 correct position (C), 4 correct letters in wrong positions, expect 0.1 + 0.05*4 = 0.3
print(reward_wordle("CRANE", "PLANT"))  # 1 correct position (A), 2 correct letters in wrong positions (N, E), expect 0.1 + 0.05*2 = 0.2
print(reward_wordle("CRANE", "ABCDE"))  # 1 correct position (E), 2 correct letters in wrong positions (C, A), expect 0.1 + 0.05*2 = 0.2
print(reward_wordle("CRANE", "12345"))  # invalid guess, expect 0.0
print(reward_wordle("CRANE", "CRAN"))   # invalid guess (length 4), expect 0.0


1.0
0.30000000000000004
0.3
0.2
0.2
0.0
0.0


----------------------
### 4. GRPO

- implement GRPO loss function


In [119]:
def print_grad_sum(model):
    for name, param in model.named_parameters():
        if param.requires_grad:
            # if grad is None, assume 0
            if param.grad is None:
                param.grad = torch.zeros_like(param)
            print(f"{name} grad: {param.grad.sum()}")
            

In [120]:
#init model
model = AutoModelForCausalLM.from_pretrained(model_name).to(device)

print_grad_sum(model)


model.embed_tokens.weight grad: 0.0
model.layers.0.self_attn.q_proj.weight grad: 0.0
model.layers.0.self_attn.k_proj.weight grad: 0.0
model.layers.0.self_attn.v_proj.weight grad: 0.0
model.layers.0.self_attn.o_proj.weight grad: 0.0
model.layers.0.self_attn.q_norm.weight grad: 0.0
model.layers.0.self_attn.k_norm.weight grad: 0.0
model.layers.0.mlp.gate_proj.weight grad: 0.0
model.layers.0.mlp.up_proj.weight grad: 0.0
model.layers.0.mlp.down_proj.weight grad: 0.0
model.layers.0.input_layernorm.weight grad: 0.0
model.layers.0.post_attention_layernorm.weight grad: 0.0
model.layers.0.pre_feedforward_layernorm.weight grad: 0.0
model.layers.0.post_feedforward_layernorm.weight grad: 0.0
model.layers.1.self_attn.q_proj.weight grad: 0.0
model.layers.1.self_attn.k_proj.weight grad: 0.0
model.layers.1.self_attn.v_proj.weight grad: 0.0
model.layers.1.self_attn.o_proj.weight grad: 0.0
model.layers.1.self_attn.q_norm.weight grad: 0.0
model.layers.1.self_attn.k_norm.weight grad: 0.0
model.layers.1.mlp

In [221]:
def get_model_prob(model, seq_ids, output_masks):
    """
    Calculate model sequence probability of OUTPUT given INPUT: P(output | input)
    
    Args:
        model: Model to use for inference
        seq_ids: Tensor of shape (B, seq_len) containing token ids of generated sequence belonging to same input
        output_masks: Tensor of shape (B, seq_len) containing mask for ouptput positions. (input and padding positions in seq_ids are set to 0)
    
    Returns:
        Tensor of shape (B,) containing probabilities for each outputsequence in batch
        
    Note:
        Uses model.forward() to get logits for sequence, then uses log_softmax to get probabilities.
        Using model.forward() helps track gradients of output tensor for backprop.
    """
    # opmask: 0 0 0 0 0 1 1 1 1 1 1     (masks of output positions, ignore padding tokens)
    # input : i n p u t o u t p u t     (model input seq)
    # output: n p u t o u t p u t -     (get logit for output tokens, ie. model probablity)
    # masks:  0 0 0 0 1 1 1 1 1 1 0     (left shift by 1 to get logits for output tokens)
    
    attention_mask = torch.ones(1, seq_ids.shape[-1])
    output = model(
        input_ids=seq_ids,
        attention_mask=attention_mask,
    )
    logits = output.logits
    log_prob = torch.log_softmax(logits, dim=-1)             # (B, seq_len, V)  -> log_prob for each token in seq
    output_indices = torch.roll(seq_ids, shifts=-1, dims=-1)        # (B, seq_len) -> left shift by 1 to get indices for output tokens
    log_prob = log_prob.gather(
        index=output_indices.unsqueeze(-1),                         # (B, seq_len, 1)  -> (B, seq_len, 1) 
        dim=-1
    )                                                               # (B, seq_len, 1)  -> log_prob for seq tokens
    log_prob = log_prob.squeeze()                                   # (B, seq_len)
    masks = torch.roll(output_masks, shifts=-1, dims=-1)            # left shift by 1
    masks[:, -1] = 0                                                # pad last token with 0 as it is not output token
    output_log_prob = log_prob * masks                              # (B, seq_len)  -> log_prob for output tokens     
    sum_log_prob = output_log_prob.sum(dim=-1)                      # (B, 1)  -> sum of log_prob for output tokens
    prob = torch.exp(sum_log_prob)                                  # (B, 1)  -> prob for output tokens
    return prob, log_prob, logits
    

def get_group_relative_reward_advantage(rewards: torch.Tensor) -> torch.Tensor:
    """
    Calculate reward advantage of sequence with "Group Relevative" reward function.
    
    Args:
        rewards: Tensor of shape (B,) containing reward values for generated sequences belonging to same input
    
    Returns:
        Tensor of shape (B,) containing reward advantage based on group relative reward function.
    """
    mean_reward = rewards.mean()
    std_reward = rewards.std()
    advantages = (rewards - mean_reward) / std_reward
    return advantages


def get_kl_divergence(prob_curr, prob_ref):
    """
    Calculate KL divergence between model and ref_model for sequence given input.

    Args:
        prob_curr: Tensor of shape (B,) containing probabilities for current model
        prob_ref: Tensor of shape (B,) containing probabilities for reference model
        seq_ids: Tensor of shape (B, seq_len) containing token ids of generated sequence belonging to same input
        output_masks: Tensor of shape (B, seq_len) containing mask for ouptput positions. (input and padding positions in seq_ids are set to 0)

    Returns:
        Tensor of shape (B,) containing KL divergence for each outputsequence in batch
    """
    kl_div = (prob_ref / prob_curr) - torch.log(prob_ref / prob_curr) - 1
    return kl_div
    


def grpo(curr_model, old_model, ref_model, seq_ids, output_masks, rewards, ep=0.01, beta=0.1):
    """
    GRPO loss function

    Args:
        curr_model: current policy model - trained model weights
        old_model: old policy model - old model weights from previous iteration
        ref_model: Reference model base model - pretrained model weights
        seq_ids: Tensor of shape (B, seq_len) containing token ids of generated sequence belonging to same input
        output_masks: Tensor of shape (B, seq_len) containing mask for ouptput positions. (input and padding positions in seq_ids are set to 0)
        rewards: Tensor of shape (B,) containing reward values for generated sequences belonging to same input
        ep: clipping threshold for reward
        beta: KL divergence regularization parameter

    Returns:
        Tensor of shape (1,) containing GRPO loss for the batch
    """
    prob_curr, _, _ = get_model_prob(curr_model, seq_ids, output_masks)
    prob_old, _, _ = get_model_prob(old_model, seq_ids, output_masks)
    prob_ref, _, _ = get_model_prob(ref_model, seq_ids, output_masks)
    adv = get_group_relative_reward_advantage(rewards)

    policy_obj = torch.min(
        (prob_curr / prob_old) * adv,
        torch.clamp(prob_curr / prob_old, 1 - ep, 1 + ep) * adv,
    )
    kl_div = get_kl_divergence(prob_curr, prob_ref)
    grpo_loss = policy_obj - (beta * kl_div)
    grpo_loss = - grpo_loss.mean()
    return grpo_loss


Test if get prod produces correct output

In [185]:
# test get_model_prob
input_len = input_pt.input_ids.shape[1]
output_masks = torch.ones(sample_gen.sequences.shape).to(device)
output_masks[:, :input_len] = 0 # mask input tokens
# TODO: mask padding tokens

p, log_prob, logits = get_model_prob(model, sample_gen.sequences, output_masks)
print(p)
print(log_prob)
print(f"Logits grad: {logits.grad}")

Logits requires grad: True
tensor([0.3907], device='mps:0', grad_fn=<ExpBackward0>)
tensor([-1.2838e+01, -1.1584e+00, -2.1250e+00, -5.1828e+00, -1.2865e+00,
        -3.1996e-02, -1.5039e-03, -6.3487e-02, -1.4876e-04, -8.3659e-01,
        -3.5763e-07, -5.0759e-03, -6.6636e-05, -5.2772e-04, -3.0656e-04,
        -3.9765e+01], device='mps:0', grad_fn=<SqueezeBackward0>)
Logits grad: None


/var/folders/g8/lf0c322561xd2t6f8mykdsxc0000gn/T/ipykernel_62523/1886068848.py:9: UserWarning: The .grad attribute of a Tensor that is not a leaf Tensor is being accessed. Its .grad attribute won't be populated during autograd.backward(). If you indeed want the .grad field to be populated for a non-leaf Tensor, use .retain_grad() on the non-leaf Tensor. If you access the non-leaf Tensor by mistake, make sure you access the leaf Tensor instead. See github.com/pytorch/pytorch/pull/30531 for more informations. (Triggered internally at /Users/runner/work/pytorch/pytorch/pytorch/build/aten/src/ATen/core/TensorBody.h:494.)
  print(f"Logits grad: {logits.grad}")


Test gradients of masked position is zero; we only want to learning from output token, not the input and padding tokens

In [186]:
def test_masked_gradients(model):
    """Test if gradients for masked positions are zero"""
    
    # Clear existing gradients
    model.zero_grad()
    model.train()
    
    # Forward pass with sequences that require gradients
    sample_input = "Tell me a joke."
    input_pt = tokenizer(sample_input, return_tensors="pt").to(device)
    input_len = input_pt.input_ids.shape[1]
    sample_gen = model.generate(
        **input_pt,
        max_new_tokens=10,
        temperature=0.5,
        return_dict_in_generate=True,
        output_scores=True,
    )
    output_masks = torch.ones(sample_gen.sequences.shape).to(device)
    output_masks[:, :input_len] = 0
    output_masks.requires_grad = True
    output_masks.retain_grad()
    sequences_with_grad = sample_gen.sequences
    p, log_prob, logits = get_model_prob(model, sequences_with_grad, output_masks)
    log_prob.retain_grad()
    logits.retain_grad()
    p.retain_grad()
    print(f"Output probability: {p}")
    print(f"log prob: {log_prob}")

    # some simple loss function like grpo
    loss = - (p / (input_len + 1) * 0.21)
    loss.retain_grad()
    print(f"loss: {loss}")

    # Grads before backward pass
    print(f"Logit grads before backward pass: {logits.grad}")
    print(f"Log_prob grads before backward pass: {log_prob.grad}")
    
    # Backward pass
    print("-"*100)
    loss.backward()
    print("Backward pass done")
    
    print(f"Loss grad: {loss.grad}")
    print(f"Prob grad: {p.grad}")
    print(f"log_prob grad: {log_prob.grad}")
    # Check if gradients for input positions are zero
    logits_grad = logits.grad
    print(f"Logits grad shape: {logits_grad.shape}")
    print(f"Logits grad: \n{logits_grad}")

    # Check input token gradients (should be zero)
    input_token_grads = logits_grad[:, :input_len-1]
    print(f"Input token grads: {input_token_grads}")
    print(f"Input token grads shape: {input_token_grads.shape}")
    print(f"Max absolute gradient for input tokens: {input_token_grads.abs().max().item():.10f}")
    print(f"Sum of input token gradients: {input_token_grads.sum().item():.10f}")
    print(f"Are all input token gradients zero? {torch.allclose(input_token_grads, torch.zeros_like(input_token_grads), atol=1e-10)}")
    
    # Check output token gradients (should be non-zero)
    output_token_grads = logits_grad[:, input_len-1:]
    print(f"Output token grads shape: {output_token_grads.shape}")
    print(f"Output token grads: {output_token_grads}")
    print(f"Max absolute gradient for output tokens: {output_token_grads.abs().max().item():.10f}")
    print(f"Number of non-zero output gradients: {(output_token_grads.abs() > 1e-10).sum().item()}")
    
    return logits_grad

# Test with your data
_ = test_masked_gradients(model)

Logits requires grad: True
Output probability: tensor([0.3907], device='mps:0', grad_fn=<ExpBackward0>)
log prob: tensor([-1.2838e+01, -1.1584e+00, -2.1250e+00, -5.1828e+00, -1.2865e+00,
        -3.1996e-02, -1.5039e-03, -6.3487e-02, -1.4876e-04, -8.3659e-01,
        -3.5763e-07, -5.0759e-03, -6.6636e-05, -5.2772e-04, -3.0656e-04,
        -3.9765e+01], device='mps:0', grad_fn=<SqueezeBackward0>)
loss: tensor([-0.0117], device='mps:0', grad_fn=<NegBackward0>)
Logit grads before backward pass: None
Log_prob grads before backward pass: None
----------------------------------------------------------------------------------------------------
Backward pass done
Loss grad: tensor([1.], device='mps:0')
Prob grad: tensor([-0.0300], device='mps:0')
log_prob grad: tensor([-0.0000, -0.0000, -0.0000, -0.0000, -0.0000, -0.0117, -0.0117, -0.0117,
        -0.0117, -0.0117, -0.0117, -0.0117, -0.0117, -0.0117, -0.0117, -0.0000],
       device='mps:0')
Logits grad shape: torch.Size([1, 16, 262144])
Logit

In [191]:
# test get_kl_divergence for same model
# test get_model_prob
sample_input = "Tell me a joke."
input_pt = tokenizer(sample_input, return_tensors="pt").to(device)
input_len = input_pt.input_ids.shape[1]
sample_gen = model.generate(
        **input_pt,
        max_new_tokens=10,
        temperature=0.5,
        return_dict_in_generate=True,
        output_scores=True,
)
input_len = input_pt.input_ids.shape[1]
output_masks = torch.ones(sample_gen.sequences.shape).to(device)
output_masks[:, :input_len] = 0 # mask input tokens

p, log_prob, logits = get_model_prob(model, sample_gen.sequences, output_masks)
print(p)

kl_div = get_kl_divergence(model, model, sample_gen.sequences, output_masks)
print(f"KLD: {get_kl_divergence(model, model, sample_gen.sequences, output_masks)}")
# Assert that KL divergence is zero when comparing same model to itself
assert kl_div == 0.0, f"KL divergence should be zero for same model, got {kl_div}"
print("✓ KL divergence is zero as expected for same model comparison")



tensor([0.2114], device='mps:0', grad_fn=<ExpBackward0>)
KLD: tensor([0.], device='mps:0', grad_fn=<SubBackward0>)
✓ KL divergence is zero as expected for same model comparison


In [ ]:
# Generate sample sequences for single input
sample_input = "Tell me a joke."
input_pt = tokenizer(sample_input, return_tensors="pt").to(device)
input_len = input_pt.input_ids.shape[1]
print(f"input_len: {input_len}")
sample_gen = model.generate(
        **input_pt,
        num_return_sequences=8,
        max_new_tokens=32,
        temperature=1,
        return_dict_in_generate=True,
)
print(f"Shape of sample_gen.sequences: {sample_gen.sequences.shape}")
input_len = input_pt.input_ids.shape[1]
print(f"sample_gen.sequences:")
print(sample_gen.sequences)

# Set input and padding tokens to zero in output_masks
output_masks = torch.ones(sample_gen.sequences.shape).to(device)
output_masks[:, :input_len] = 0 # mask input tokens
pad_token_id = tokenizer.pad_token_id
print(f"pad_token_id: {pad_token_id}")
if pad_token_id is not None:
    padding_mask = (sample_gen.sequences == pad_token_id)
    output_masks[padding_mask] = 0
print(f"output_masks:")
print(output_masks)

# decode the sequences
for i in range(sample_gen.sequences.shape[0]):
    print(i, tokenizer.decode(sample_gen.sequences[i], skip_special_tokens=True))

input_len: 6
Shape of sample_gen.sequences: torch.Size([8, 38])
sample_gen.sequences:
tensor([[     2,  54593,    786,    496,  31481, 236761,    108,  11355,   1602,
            506,  30979,   3798,   1024, 236881,    108,  17574,    625,    691,
           1156,  20718, 236888,    108,   7243,    108,  38786,    611,   1133,
            786,    531,   3442,    611,   2264,    886, 236881,    106,      0,
              0,      0],
        [     2,  54593,    786,    496,  31481, 236761,    108,  11355,   1602,
            506,  30979,   3798,   1024, 236881,    108,  17574,    625,    691,
           1156,  20718, 236888,    108,   7243,    108,  38786,    611,   1133,
            786,    531,   3442,    611,   2264,    886, 236881,    106,      0,
              0,      0],
        [     2,  54593,    786,    496,  31481, 236761,    108,  11355,   1602,
            506,  30979,   3798,   1024, 236881,    108,  17574,    625,    691,
           1156,  20718, 236888,    108,   7243,    

In [228]:
# test grpo loss function
# rewards = torch.randn(8).to(device)
print("⚠️ When current, old and ref models are same 👉 GRPO loss should be equal to negative mean of advantages")
adv = get_group_relative_reward_advantage(rewards)
print(f"rewards: {rewards}")
print(f"adv: {adv}")
grpo_loss = grpo(model, model, model, sample_gen.sequences, output_masks, rewards)
# assert
assert grpo_loss == -adv.mean(), f"GRPO loss should be equal to mean of advantages, got {grpo_loss} and {adv.mean()}"
assert abs(grpo_loss) < 1e-6, f"GRPO loss should be approximately zero, got {grpo_loss}"
print(f"GRPO loss: {grpo_loss:.6f}")
print(f"Negative mean of advantages: {-adv.mean():.6f}")
print("✅ GRPO loss is equal to negative mean of advantages")
print("✅ GRPO loss is close to zero")

⚠️ When current, old and ref models are same 👉 GRPO loss should be equal to negative mean of advantages
rewards: tensor([-0.0835, -0.2934,  0.9346, -1.5604,  0.5261, -0.8268, -0.6088,  0.1613],
       device='mps:0')
adv: tensor([ 0.1712, -0.0943,  1.4590, -1.6969,  0.9423, -0.7690, -0.4932,  0.4809],
       device='mps:0')
GRPO loss: -0.000000
Negative mean of advantages: -0.000000
✅ GRPO loss is equal to negative mean of advantages
✅ GRPO loss is close to zero


------------------
### 5. Training Loop

GRPO training loop
```
for sample in dataset
    1.  generations -> model.generate(sample, n=8)
        batch generations togethers with padding
        create output token mask, 
    2.  reward, advantage --> score (generations)
        loss -> grpo; seq_prob; kl
    3.  backprop -> loss.backward() 
        optimizer.step()
```

In [124]:
import torch.optim as optim

adam = optim.Adam(model.parameters())
adam.zero_grad()

In [125]:
model.eval()
adam.zero_grad()

sample_input = "Tell me a joke."
input_pt = tokenizer(sample_input, return_tensors="pt").to(device)

# sample generation (auto-regressive)
sample_gen = model.generate(
    **input_pt,
    max_new_tokens=10,
    temperature=0.5,
    return_dict_in_generate=True,
    output_scores=True,
)
print(f"Output sequence (len:{sample_gen.sequences[0].shape[0]})--> {tokenizer.decode(sample_gen.sequences[0], skip_special_tokens=True)}")
input_len = input_pt.input_ids.shape[1]
print(f"input_len: {input_len}")
seq_len = sample_gen.sequences[0].shape[0]
gen_seq = sample_gen.sequences[0][input_len:]
print(f"Generated sequence (len:{gen_seq.shape[0]})--> {tokenizer.decode(gen_seq, skip_special_tokens=True)}")

# model forward pass - to get logits for the generated sequence
attention_mask = torch.ones(1, seq_len)
output_logits = model(
    input_ids=sample_gen.sequences,
    attention_mask=attention_mask,
)
# sum of log_probs for ids in sample_gen.sequences[0]
gen_seq_logits = output_logits.logits[:,input_len-1:-1,:]   # ***IMP*** logits for generated sequence input sequence left shifted by 1
gen_seq_scores = torch.log_softmax(gen_seq_logits, dim=-1)
gen_seq_ids = sample_gen.sequences.unsqueeze(-1)[:,input_len:,:]    # ids of generated sequence
gen_seq_log_prob = gen_seq_scores.gather(dim=2, index=gen_seq_ids).squeeze()

sum_gen_seq_log_prob = gen_seq_log_prob.sum()
gen_seq_prob = torch.exp(sum_gen_seq_log_prob)
print(f"sum_gen_seq_log_prob: {sum_gen_seq_log_prob:.4}")
print(f"gen_seq_prob: {gen_seq_prob:.2%}")
# print prob of each token in gen_seq
for i in range(gen_seq.shape[0]):
    print(f"\"{tokenizer.decode(gen_seq[i])}\"\t| token {gen_seq[i]} prob: {torch.exp(gen_seq_log_prob[i]):.2%} | log prob: {gen_seq_log_prob[i]:.4}")
gen_seq_prob


Output sequence (len:16)--> Tell me a joke.

Why did the scarecrow win an award?
input_len: 6
Generated sequence (len:10)--> 

Why did the scarecrow win an award?
sum_gen_seq_log_prob: -0.9397
gen_seq_prob: 39.07%
"

"	| token 108 prob: 96.85% | log prob: -0.032
"Why"	| token 11355 prob: 99.85% | log prob: -0.001504
" did"	| token 1602 prob: 93.85% | log prob: -0.06349
" the"	| token 506 prob: 99.99% | log prob: -0.0001488
" scare"	| token 55134 prob: 43.32% | log prob: -0.8366
"crow"	| token 47129 prob: 100.00% | log prob: -3.576e-07
" win"	| token 3345 prob: 99.49% | log prob: -0.005076
" an"	| token 614 prob: 99.99% | log prob: -6.664e-05
" award"	| token 8054 prob: 99.95% | log prob: -0.0005277
"?"	| token 236881 prob: 99.97% | log prob: -0.0003066


tensor(0.3907, device='mps:0', grad_fn=<ExpBackward0>)

"

"	| token 108 prob: 100.00%
"Why"	| token 11355 prob: 100.00%
" did"	| token 1602 prob: 100.00%
" the"	| token 506 prob: 100.00%
" bicycle"	| token 30979 prob: 35.67%
" fall"	| token 3798 prob: 100.00%
" over"	| token 1024 prob: 100.00%
"?"	| token 236881 prob: 100.00%
"

"	| token 108 prob: 94.77%
"Because"	| token 17574 prob: 100.00%
Sequence 1 log prob: -1.085
Sequence 1 prob: 33.80%
Sequence 1 string: 

Why did the bicycle fall over?

Because

In [126]:
# calculate loss
loss = - (gen_seq_prob/0.1 * 1)
print(f"loss: {loss:.4}")

# calculate gradient
# loss.backward()

# print grad sum
print_grad_sum(model)



loss: -3.907
model.embed_tokens.weight grad: 0.0
model.layers.0.self_attn.q_proj.weight grad: 0.0
model.layers.0.self_attn.k_proj.weight grad: 0.0
model.layers.0.self_attn.v_proj.weight grad: 0.0
model.layers.0.self_attn.o_proj.weight grad: 0.0
model.layers.0.self_attn.q_norm.weight grad: 0.0
model.layers.0.self_attn.k_norm.weight grad: 0.0
model.layers.0.mlp.gate_proj.weight grad: 0.0
model.layers.0.mlp.up_proj.weight grad: 0.0
model.layers.0.mlp.down_proj.weight grad: 0.0
model.layers.0.input_layernorm.weight grad: 0.0
model.layers.0.post_attention_layernorm.weight grad: 0.0
model.layers.0.pre_feedforward_layernorm.weight grad: 0.0
model.layers.0.post_feedforward_layernorm.weight grad: 0.0
model.layers.1.self_attn.q_proj.weight grad: 0.0
model.layers.1.self_attn.k_proj.weight grad: 0.0
model.layers.1.self_attn.v_proj.weight grad: 0.0
model.layers.1.self_attn.o_proj.weight grad: 0.0
model.layers.1.self_attn.q_norm.weight grad: 0.0
model.layers.1.self_attn.k_norm.weight grad: 0.0
model